### PR TITLE
We probably need to refresh venvs with each release, how to do that properly?

### PR DESCRIPTION
**Pull request by udislau**

### CODE REVIEW

File: backend/aiconsole/core/code_running/virtual_env/download_python.py [hunk 1]
```diff
@@ -2,16 +2,13 @@
...
+import urllib.request
 
 _log = logging.getLogger(__name__)
 logging.basicConfig(level=logging.INFO)
 
-PYTHON_VERSION = "3.10.13+20231002"
-
 
-def download_python():
+def download_python(python_version, tag):
```
Comment (cielecki) [inline backend/aiconsole/core/code_running/virtual_env/download_python.py:11]: A minor issue: date / date_str will probably be a better name than tag

File: backend/aiconsole/core/code_running/code_interpreters/languages/python.py [hunk 2]
```diff
@@ -71,9 +72,27 @@ def preprocess_python(code: str, materials: list[Material]):
...
-    code = "\n".join(
-        [f"import os; os.system({line[1:]!r})" if line.startswith("!") else line for line in code.split("\n")]
-    )
+    # code = "\n".join(
+    #     [f"import os; os.system({line[1:]!r})" if line.startswith("!") else line for line in code.split("\n")]
+    # )
+
+    updated_code = []
+    for line in code.split("\n"):
+        if line.startswith("!pip install") or line.startswith("!pip3 install"):
```
[the remark this answers is not in the document]
Comment (udislau) [reply backend/aiconsole/core/code_running/code_interpreters/languages/python.py]: Removed

File: backend/aiconsole/core/code_running/virtual_env/download_python.py (hunk 1 above)
Comment (udislau) [reply backend/aiconsole/core/code_running/virtual_env/download_python.py]: resolved

File: backend/aiconsole/core/code_running/virtual_env/create_dedicated_venv.py [hunk 3]
```diff
@@ -1,29 +1,78 @@
...
     else:
         _log.info(f"Venv already exists in {venv_path}")
 
+        venv_python_executable = str(get_current_project_venv_python_path())
+
+        venv_python_version = get_python_version(venv_python_executable)
+        if venv_python_version:
+            _log.info(f"Valid venv python executable")
+        else:
+            _log.info(f"Reconstructing venv due to the different Python version({system_python_version})")
```
[the remark this answers is not in the document]
Comment (udislau) [reply backend/aiconsole/core/code_running/virtual_env/create_dedicated_venv.py]: These are inherited from bundled Python, so always up-to-date.

**Comment by udislau** [comment]
@cielecki Looking for approval 🙂

**Review by cielecki — APPROVED** [review]

**Pull request merged**